### PR TITLE
transformations: New linalg-fuse-multiply-add pass

### PR DESCRIPTION
**Pull request by n-io**
Introduces a new pass to fuse `linalg.mul` and `linalg.add` into `linalg.generic` with the functionality of an FMA op.

### CODE REVIEW

**Comment by n-io** [comment]
This is a replacement for #3344 @superlopuh @AntonLydike

File: xdsl/transforms/linalg_transformations.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            add_operand = (
+                add.inputs[0] if mul.res[0] == add.inputs[1] else add.inputs[1]
+            )
+
+            # # prefer dest that is not used as an input
+            # dest = (
+            #     mul.outputs[0]
+            #     if mul.outputs[0] not in [*mul.inputs, *add.inputs]
+            #     else add.outputs[0]
+            # )
```
Comment (superlopuh) [inline xdsl/transforms/linalg_transformations.py:80]: commented code

File: xdsl/transforms/linalg_transformations.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        inputs,
+        outputs,
+        body,
+        4 * [AffineMapAttr(AffineMap.from_callable(lambda i,: (i,)))],
+        [linalg.IteratorTypeAttr.parallel()],
+        [out.type],
+    )
+
+
+@dataclass
```
Comment (superlopuh) [inline xdsl/transforms/linalg_transformations.py:42]: nit:
```suggestion
@dataclass(frozen=True)
```

**Review by superlopuh — APPROVED** [review]
The code looks good, although my understanding is that linalg is not designed to be used in this way, and instead fuses after tiling. I wonder if there's a similar direction for you to look into, maybe for later.

File: xdsl/transforms/linalg_transformations.py (hunk 2 above)
Comment (n-io) [reply xdsl/transforms/linalg_transformations.py]: Done as suggested.

File: xdsl/transforms/linalg_transformations.py (hunk 1 above)
Comment (n-io) [reply xdsl/transforms/linalg_transformations.py]: Whoops, removed.

**Comment by n-io** [comment]
It made sense in my mind to do this sort of op fusion before bufferization, though maybe it doesn't make a difference?

**Pull request merged**